overlord/snapstate/handlers.go

```diff
@@ -470,6 +524,10 @@ func (m *SnapManager) doLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
+	if err := m.createSnapContext(st, snapsup.Name()); err != nil {
```

Comment:
Instead of using EnsureState above, that can be a more trivial atomic write and this call can be made conditional on there being no other revision of this snap at the moment (snapst.Current).

Also, this call should be closer to the end of the function, after LinkSnap itself has succeeded, otherwise we'll write down the file and keep leftover state around in case something else here fails. Talk to @pedronis for details on this.